controller/object_store_controller.go

```diff
@@ -0,0 +1,1159 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+var (
+	OneHour, _ = time.ParseDuration("1h")
+)
+
+type ObjectStoreController struct {
+	*baseController
+
+	controllerID string
+
+	namespace string
+	ds        *datastore.DataStore
+	s3gwImage string
+	uiImage   string
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewObjectStoreController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string,
+	objectStoreImage string,
+	objectStoreUIImage string,
+) *ObjectStoreController {
+	osc := &ObjectStoreController{
+		baseController: newBaseController("object-store", logger),
+		controllerID:   controllerID,
+		namespace:      namespace,
+		ds:             ds,
+		s3gwImage:      objectStoreImage,
+		uiImage:        objectStoreUIImage,
+	}
+
+	ds.ObjectStoreInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    osc.enqueueObjectStore,
+			UpdateFunc: func(old, cur interface{}) { osc.enqueueObjectStore(cur) },
+			DeleteFunc: osc.enqueueObjectStore,
+		},
+		OneHour,
+	)
+
+	ds.DeploymentInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    osc.enqueueDeployment,
+			UpdateFunc: func(old, cur interface{}) { osc.enqueueDeployment(cur) },
+			DeleteFunc: osc.enqueueDeployment,
+		},
+		OneHour,
```

Comment:
We don't need resync for this informer because we already have resync for the main informer `ObjectStoreInformer`